bcachefs/bcachefsmodule.c

```diff
@@ -272,20 +272,24 @@ static PyObject *PyBcachefs_iterator_next(PyBcachefs_iterator *self)
     const Bcachefs *fs = &self->_pyfs->_fs;
     Bcachefs_iterator *iter = self->_iter;
     const struct bch_val *bch_val = Bcachefs_iter_next(fs, iter);
+    while (iter->next_it)
+    {
+        iter = iter->next_it;
+    }
     if (bch_val && iter->type == BTREE_ID_extents)
     {
         Bcachefs_extent extent = Bcachefs_iter_make_extent(fs, iter);
-        return Py_BuildValue("KKKK", extent.inode, extent.file_offset, extent.offset, extent.size);
+        return Py_BuildValue("KKKKK", extent.inode, extent.file_offset, extent.offset, extent.size, (uint64_t)&iter);
```

Comment:
`&iter` takes the address of the variable `iter` *on the stack and returns it*. As soon as this function returns, that address is invalid.

You might have intended to return `(uintptr_t)iter` (no `&`) here and in the other two uses below. However, that's still fishy unless you have guarantees on the lifetime of these objects. If you don't, things become hairy fast.

If `Bcachefs_iterator` had been a Python object, it would have been ideal to return it as a Python object (`Py_BuildValue` signature `"O"`) and thus reference-count it, but unfortunately it isn't a Python object.

Reply:
True what was I thinking! `(uintptr_t)iter` is what is necessary here.

I'm less interested into the object then the pointer. My need here is to return an id to an iterator to know when the iteration over its `bsets` is done, and thus clean the cache in Python for the next set of `bsets`.